### PR TITLE
fix bug in TuneWrapper for param in used in predict function, issue 2472

### PR DESCRIPTION
**Pull request by berndbischl**
fixes issue #2472 

unit tests are missing. please help and merge

### CODE REVIEW

File: R/TuneWrapper.R [hunk 1]
```diff
@@ -75,7 +75,10 @@ trainLearner.TuneWrapper = function(.learner, .task, .subset = NULL,  ...) {
 #' @export
 predictLearner.TuneWrapper = function(.learner, .model, .newdata, ...) {
   lrn = setHyperPars(.learner$next.learner, par.vals = .model$learner.model$opt.result$x)
-  predictLearner(lrn, .model$learner.model$next.model, .newdata, ...)
+  arglist = list(.learner = lrn, .model = .model$learner.model$next.model, .newdata = .newdata)
+  arglist = insert(arglist, list(...))
+  arglist = insert(arglist, .model$learner.model$opt.result$x)
```
Comment (mb706) [inline R/TuneWrapper.R]: Should probably only insert the parameters that have `when %in% c("predict", "both")`.

Maybe one could just copy [the lines from predict.R](https://github.com/mlr-org/mlr/blob/e547bae5e776c9378a764e33962f82cb44f95ecd/R/predict.R#L101-L106), if we are sure there are no other wrappers that introduce a discrepancy between the `par.vals` of a learner and the `...` parameters on purpose.

**Comment by jakob-r** [comment]
Some notes regarding my observation:

- `setHyperPars` within `predictLearner` is without effect because they will never be requested. The hyperparams got extracted in `predict.(WrappedModel)` and are put into `...`. 
- We don't know if the `par.vals` got changed after the training of the TuneWrapped Learner. 
- We will always overwrite the existing par.vals with the ones from the learner.
- We assume that the user does not change par.vals and between training and predict. This holds especially as in most cases both is done internally in one step.

**Comment by mb706** [comment]
> setHyperPars within predictLearner is without effect because they will never be requested. The hyperparams got extracted in predict.(WrappedModel) and are put into ....

we should probably not rely on this; right now the parameter values are given both inside the `...` and as part of the `.learner` object; its probably good to make sure they agree.

> We don't know if the par.vals got changed after the training of the TuneWrapped Learner.

How would that happen (assuming the user does not directly modify the `model$learner` object, in which case, too bad)?

> We will always overwrite the existing par.vals with the ones from the learner.

You mean the ones being tuned over? It might be worth considering to *remove* the parameters found in the tuning param set from the wrapped learner's param set, so the user gets an error message when he tries to set a param that will be overwritten later.

>We assume that the user does not change par.vals and between training and predict. This holds especially as in most cases both is done internally in one step.

I'm confused again, how would that happen? Does any part of `mlr` support that?

**Comment by jakob-r** [comment]
> > setHyperPars within predictLearner is without effect because they will never be requested. The hyperparams got extracted in predict.(WrappedModel) and are put into ....
> 
> we should probably not rely on this; right now the parameter values are given both inside the `...` and as part of the `.learner` object; its probably good to make sure they agree.

That is the architecture of mlr at the moment and we are relying on it everywhere. In `train` and `predict` the parameters are read and put into the arguments.
This is probably not the best design but keeping the arguments and the par.vals "in sync" would only be a lot of effort.

> > We don't know if the par.vals got changed after the training of the TuneWrapped Learner.
> 
> How would that happen (assuming the user does not directly modify the `model$learner` object, in which case, too bad)?

Like that, yes., ? `mod = train(...); setHyperPars(mod$learner, ...); predict(mod, ...)`

> > We will always overwrite the existing par.vals with the ones from the learner.
> 
> You mean the ones being tuned over? It might be worth considering to _remove_ the parameters found in the tuning param set from the wrapped learner's param set, so the user gets an error message when he tries to set a param that will be overwritten later.

A warning somewhere might be useful. But I am afraid we would generate warnings for cases where we have set the par.vals in the definition of the learner. This can entail a lot of work.

> > We assume that the user does not change par.vals and between training and predict. This holds especially as in most cases both is done internally in one step.
> 
> I'm confused again, how would that happen? Does any part of `mlr` support that?
See above.

My suggestion: We should merge this PR asap and can add convenience later.

**Comment by mb706** [comment]
> That is the architecture of mlr at the moment and we are relying on it everywhere. In train and predict the parameters are read and put into the arguments.

Do you consider [this](https://github.com/mlr-org/mlr/blob/f2817efb5d033adf2d090cd435852c974f1722d1/R/RLearner_classif_gamboost.R#L65) to be a bug?

>`mod = train(...); setHyperPars(mod$learner, ...); predict(mod, ...)`

has "`model$learner <- [something]`" ever been documented as supported behaviour? If the user modifies internals of objects then of course he can break things without bounds, I thought of `mod$learner` as such an internal.

**Comment by berndbischl** [comment]
i do not understand the unit test. can you please add at least some comments? 
how the test is setup and why you expect certain things?

**Comment by berndbischl** [comment]
also please note: a PR for fixing a certain bug is probably not the best place to dicuss general architecture questions. please move this to a clean issue if you want to discuss this...

**Comment by jakob-r** [comment]
> Do you consider [this](https://github.com/mlr-org/mlr/blob/f2817efb5d033adf2d090cd435852c974f1722d1/R/RLearner_classif_gamboost.R#L65) to be a bug?

I would not call it a bug as it's just affecting some error message. But the whole code is strange. How do we know that it's really the wrong family that leads to NAs? But this is another issue and should not be discussed here.

> i do not understand the unit test.

Better now?

**Comment by jakob-r** [comment]
@berndbischl *ping*

File: R/TuneWrapper.R [hunk 2]
```diff
@@ -74,8 +74,16 @@ trainLearner.TuneWrapper = function(.learner, .task, .subset = NULL,  ...) {
 
 #' @export
 predictLearner.TuneWrapper = function(.learner, .model, .newdata, ...) {
-  lrn = setHyperPars(.learner$next.learner, par.vals = .model$learner.model$opt.result$x)
```
Comment (mb706) [inline R/TuneWrapper.R]: FYI I am relatively sure that removing this line (and not putting `.learner = lrn` further down etc) will break things.

Comment (jakob-r) [reply R/TuneWrapper.R]: you mean `.learner$next.learner`? That moved to another line. But I added setHyperPars now again.

**Comment by pat-s** [comment]
@jakob-r If the build passes, can this be merged? I would add an entry to NEWS.md

**Pull request merged**